### ISSUE TITLE
Extention of FSDP trainer to multi-node settings

### ISSUE THREAD

**Issue by sapkotaruz11** [dice-group/dice-embeddings#422]
With the current FSDP trainer, we can train large-scale knowledge graph embeddings on a single node with multiple GPUs — for example, a KG with roughly 200M entities and 1B triples on 4×H100  can be trained for 500 epochs in around 53 hours with an embedding dimension of 256. Extending FSDP to a multi-node setting would let us scale to even larger graphs. To do this, both the embedding tables and the optimiser states must be fully sharded across nodes, rather than each node handling its own copies of the embedding tables and optimiser states.

**Comment by Demirrr** [comment]
@sapkotaruz11 please report the details about this example 

> KG with roughly 200M entities and 1B triples on 4×H100 can be trained for 500 epochs in around 53 hours with an embedding dimension of 256

Ideally, please share the config json file as well as the epoch losses please

**Comment by sapkotaruz11** [comment]
@sshivam95 can you provide the details?

**Comment by sshivam95** [comment]
Here is the [configuration.json](https://github.com/user-attachments/files/30266381/configuration.json) file. It was not able to complete a single epoch therefore, I am attaching the log file [2031759.log](https://github.com/user-attachments/files/30266417/2031759.log). The format of the log file might be a bit confusing because it is a slurm log but you can see the `loss_step` and `loss_epoch` of each batch.

**Comment by Demirrr** [comment]
I’m a bit confused by the findings from this large-scale experiment, as these two statements seem to contradict each other:

@sapkotaruz11

a KG with roughly 200M entities and 1B triples on 4×H100 can be trained for 500 epochs in around 53 hours with an embedding dimension of 256

@sshivam95

It was not able to complete a single epoch therefore, I am attaching the log file [2031759.log](https://github.com/user-attachments/files/30266417/2031759.log).

Could you help clarify this discrepancy? Am I missing a difference in configuration between the two runs?

**Comment by sshivam95** [comment]
@Demirrr The log file of the KG that I have sent have 396400379 entites which is failing. The one which @sapkotaruz11 is asking are as follows: 

[configuration.json](https://github.com/user-attachments/files/30267661/configuration.json)

[epoch_losses.csv](https://github.com/user-attachments/files/30267665/epoch_losses.csv)

**Comment by Demirrr** [comment]
Few details about the issue: — Extension of FSDP trainer to multi-node settings

## Evidence reviewed

Two experiments were shared on the issue to back the original 200M-entity/1B-triple/53h/4×H100
benchmark:

| | "Succeeding" run (matches issue's 53h claim) | "Failing" run |
|---|---|---|
| KG | DBpedia (`dbpedia-latest-merged-clean.nt`, ~200M entities) | Wikidata (`latest-truthy-clean.nt`, 396,400,379 entities) |
| `batch_size` | 500,000 | 1,000 |
| `fsdp_trainer_kwargs.precision` | `float32` | `float16` |
| `fsdp_trainer_kwargs.fsdp_optim_device` | `gpu` | `cpu` (offloaded) |
| Result | 500/500 epochs completed; `EpochLoss` `0.505 → 0.0166`, smooth monotonic decrease, no NaNs | Killed at epoch 1, batch **4,979/349,410** (~1.4% into the first epoch) |

The failing run's SLURM log:

```
Signal 9 (SIGKILL) received by PID 21382
[2026-07-21T12:57:52.883] error: Detected 2 oom_kill events in StepId=2031759.0. Some of the step tasks have been OOM Killed.
srun: error: gpu1006: task 0: Out Of Memory
```

Single host (`gpu1006`), killed almost immediately into epoch 1 — a genuine host-RAM OOM

## Analysis

**The DBpedia benchmark is verified.** The 53h/500-epoch/~200M-entity number .

**The Wikidata OOM — it's a memory-capacity wall.** At 396M entities × 256
dims ≈ 101B embedding parameters, Adam's two fp32 moment buffers alone are on the order of
~800GB, plus the embedding table itself — over 1TB total. `fsdp_optim_device=cpu` was set
specifically because this doesn't fit in GPU HBM; it doesn't fit in node RAM either. Spreading
that state across more local ranks on the *same* node doesn't reduce the total memory demand —
only more nodes (more aggregate RAM) does. So somewhere between 200M and 396M entities, a
single 4×H100 node's combined GPU+CPU memory is already the ceiling.

This directly validates the issue's premise: multi-node isn't a "nice to have for going even
bigger," it's the only way to cross that memory ceiling at all.

## Technical read on the ask itself

1. **Less new work than "extend FSDP to multi-node" sounds like.** The row-wise entity
   sharding already goes through `dist.all_to_all_single`, and dense params go through FSDP —
   both are topology-agnostic `torch.distributed`/NCCL primitives. Multi-node may mostly be a
   launch/config problem (`torchrun --nnodes`, `--rdzv_endpoint`, NCCL env vars for the
   interconnect) rather than a rewrite of the sharding logic.

2. **Real risk is network cost, not correctness.** Every batch does an `all_to_all` to route
   entity-index lookups to their owning rank (`fsdp_models.py:255-288`). Within a node that
   rides NVLink; across nodes it's bounded by Ethernet/IB bandwidth — an order of magnitude+
   slower. At 200M+ entities this could dominate step time. A naive multi-node extension (same
   all-to-all over more, slower links) likely won't scale the way the issue implies; a
   hierarchical scheme (shard within-node first, then across nodes) would probably be needed.

3. **The end-of-training gather doesn't scale with the stated goal.** `_materialize_model` /
   `_gather_shards` (`fsdp_models.py:376-428`) collects the entire entity table onto rank 0's
   CPU memory at the end of training. The Wikidata OOM shows a ~396M-entity table alone is
   already beyond one node's RAM — so "gather everything onto one node" becomes the actual
   scaling bottleneck for exactly the KG sizes this feature targets. This needs to become "each
   rank writes its own shard to disk" instead.

4. **Mid-run fault tolerance is missing.** Periodic checkpointing
   (`KeepLastModelCallback`/evaluation callback in `callbacks.py`) only fires on
   `is_global_zero` and saves `self.raw_model`, which under FSDP is just rank 0's local entity
   shard, not the full table. Already a latent bug for single-node long runs; multi-node makes
   it worse — more nodes over more hours means node failure becomes a real possibility, with
   currently no way to resume a partially-completed sharded run.

## Sequence of Actions:

1. **Verified** — the single-node benchmark is real and reproducible (done, via this review).
2. **Fix sharded checkpoint save/resume** — needed even for the current single-node case, and a
   prerequisite for any multi-node fault tolerance.
3. **Then tackle multi-node** with a topology-aware (intra-node vs inter-node) communication
   scheme rather than a flat extension of the current all-to-all, or the network cost is likely
   to eat the scaling benefit.

Suggested next step on the issue: post that the discrepancy is resolved (two different KGs,
not a contradiction), and that the Wikidata OOM log is itself the strongest evidence for why
this feature is needed.

**Comment by sshivam95** [comment]
With the new commits, when I try to run on multi-node, multi-gpu setup, I am not able to get the print logs (e.g., description of the dataset, time taken to read the KG etc.). This is the command I am running:

```
#!/bin/sh
#SBATCH -J 100K_Wikidata_Keci_FSDP
#SBATCH -N 2
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=192
#SBATCH -t 07-00:00:00
#SBATCH -A hpc-prf-whale
#SBATCH --gres=gpu:h100:4
#SBATCH --partition=gpu_h100
#SBATCH -e dice-embeddings/KGs/UKGe/Wikidata/logs/%x/%j.log
#SBATCH -o dice-embeddings/KGs/UKGe/Wikidata/logs/%x/%j.log

module reset
module load lang/Miniforge3/24.1.2-0
source $(conda info --base)/etc/profile.d/conda.sh
conda deactivate
conda activate dice
module list

JOB_ID=$SLURM_JOB_ID
data_path="dice-embeddings/KGs/UKGe/Wikidata/latest-truthy-clean.nt"

file_name=$(basename "$data_path")
dataset_name="${file_name%.*}"
run_output_dir="dice-embeddings/KGs/UKGe/Wikidata/models/${dataset_name}_100_000"
mkdir -p "$run_output_dir"

# Keep preprocessing parallel, but cap per-rank thread pools and dataloader workers.
NNODES=${SLURM_JOB_NUM_NODES:-${SLURM_NNODES:-2}}
GPUS_PER_NODE=${SLURM_GPUS_ON_NODE:-4}
TOTAL_CPUS_PER_NODE=8
CPU_CORES_PER_RANK=$((TOTAL_CPUS_PER_NODE / GPUS_PER_NODE))
if [ "$CPU_CORES_PER_RANK" -lt 1 ]; then
    CPU_CORES_PER_RANK=1
fi

DL_WORKERS=${DL_WORKERS:-8}
PREFETCH_FACTOR=${PREFETCH_FACTOR:-2}
BATCH_SIZE=${BATCH_SIZE:-500000}
NUM_EPOCHS=${NUM_EPOCHS:-100}
EMBEDDING_DIM=${EMBEDDING_DIM:-256}
CHECKPOINT_EVERY_N_EPOCHS=${CHECKPOINT_EVERY_N_EPOCHS:-1}
MASTER_ADDR=${MASTER_ADDR:-$(scontrol show hostnames "$SLURM_JOB_NODELIST" | head -n 1)}
MASTER_PORT=${MASTER_PORT:-$((29000 + (SLURM_JOB_ID % 1000)))}
RDZV_ID=${RDZV_ID:-$SLURM_JOB_ID}

export OMP_NUM_THREADS=$CPU_CORES_PER_RANK
export MKL_NUM_THREADS=$CPU_CORES_PER_RANK
export OPENBLAS_NUM_THREADS=$CPU_CORES_PER_RANK
export NUMBA_NUM_THREADS=$CPU_CORES_PER_RANK
export OMP_PLACES=cores
export OMP_PROC_BIND=false
export PYTHONUNBUFFERED=true
export NCCL_ASYNC_ERROR_HANDLING=1
export TORCH_NCCL_BLOCKING_WAIT=1
export TORCH_DISTRIBUTED_DEBUG=DETAIL

FSDP_KWARGS=$(printf '{"precision":"bfloat16","fsdp_optim_device":"cpu","num_workers":%s,"prefetch_factor":%s,"checkpoint_every_n_epochs":%s}' \
    "$DL_WORKERS" "$PREFETCH_FACTOR" "$CHECKPOINT_EVERY_N_EPOCHS")

echo "Launching torchFSDP on ${NNODES} nodes x ${GPUS_PER_NODE} GPUs per node"
echo "MASTER_ADDR=${MASTER_ADDR} MASTER_PORT=${MASTER_PORT} RDZV_ID=${RDZV_ID}"
echo "BATCH_SIZE=${BATCH_SIZE} DL_WORKERS=${DL_WORKERS} CPU_CORES_PER_RANK=${CPU_CORES_PER_RANK}"

srun --cpu-bind=none \
    torchrun \
    --nnodes="$NNODES" \
    --nproc-per-node="$GPUS_PER_NODE" \
    --node_rank="$SLURM_NODEID" \
    --rdzv_backend c10d \
    --rdzv_endpoint "${MASTER_ADDR}:${MASTER_PORT}" \
    --rdzv_id "$RDZV_ID" \
    dicee --path_single_kg "$data_path" \
    --path_to_store_single_run "$run_output_dir" \
        --model Keci \
        --num_epochs "$NUM_EPOCHS" \
        --p 0 --q 1 \
        --embedding_dim "$EMBEDDING_DIM" \
        --scoring_technique NegSample \
        --eval_model None \
        --batch_size "$BATCH_SIZE" \
        --optim Adam \
        --trainer torchFSDP \
        --num_core "$TOTAL_CPUS_PER_NODE" \
        --fsdp_trainer_kwargs "$FSDP_KWARGS" \
        --reuse_existing_run_dir

echo "DONE"
conda deactivate
exit 0
```

The same happens with PL and DDP trainer as well. Their Print callbacks are now off somehow. The same is happening on single node multi-GPU setup as well. Previously, that was not the case.

**Comment by sapkotaruz11** [comment]
Printing the info is now handled via logging. You can set the logging level via the CLI.

**Comment by Demirrr** [comment]
Thank you @sapkotaruz11 . @sshivam95 here is how to use the loging level via the CLI https://github.com/dice-group/dice-embeddings#logging

**Comment by Demirrr** [comment]
@sshivam95 please share as much as info possible about this large scale experiment. I am quite curious about the details :)

**Comment by sshivam95** [comment]
> With the current FSDP trainer, we can train large-scale knowledge graph embeddings on a single node with multiple GPUs — for example, a KG with roughly 200M entities and 1B triples on 4×H100 can be trained for 500 epochs in around 53 hours with an embedding dimension of 256. Extending FSDP to a multi-node setting would let us scale to even larger graphs. To do this, both the embedding tables and the optimiser states must be fully sharded across nodes, rather than each node handling its own copies of the embedding tables and optimiser states.

A slight correction in this stat, FSDP trainer was able to run the [DBpedia snapshot version 2022-12](https://files.dice-research.org/datasets/DBPedia/dbpedia-generic-snapshot-2022-12/), which has the following description:
```
Number of entities: 156,456,361
Number of relations: 13,943
Number of triples on train set: 843,684,312
Number of triples on valid set: 0
Number of triples on test set: 0
Entity Index: 20.66265 GB
Relation Index: 0.00148 GB
```
, in 59.225 hours for 500 epochs on 256 dimension. Here the batch size was $500,000$. Here are the report, configuration and loss epoch files of that run.

[report.json](https://github.com/user-attachments/files/30548249/report.json)
[epoch_losses.csv](https://github.com/user-attachments/files/30548248/epoch_losses.csv)
[configuration.json](https://github.com/user-attachments/files/30548250/configuration.json)

**Comment by sshivam95** [comment]
The training on the wikidata KG with 1,397,639,131 triples and 396,400,379 entities with 256 dimension for 100 epochs took 55 hrs and 30 min. However, after the training was complete, it gave an oom error on the cpu side. I was using 2 nodes with 730GB usable memory. After checking the memory usage of the slurm job, after the training both nodes consumed 752 and 767 GB respectively and the job failed. Here is the metric table.

| Node | **mem_used (GB)** | | | **nv_mem_util (%)** | | | | **nv_util (%)** | | | | **nv_fb_mem_used (GB)** | | | |
|:-----|------------------:|------------------:|------------------:|:---------------------|----------------:|----------------:|----------------:|:-----------------|------------:|------------:|------------:|:-----------------------|------------:|------------:|------------:|
|      | **Min** | **Avg** | **Max** | **ID** | **Min** | **Avg** | **Max** | **ID** | **Min** | **Avg** | **Max** | **ID** | **Min** | **Avg** | **Max** |
| gpu1001 | 548.49 | 607.61 | 752.07 | 00000000:06:00.0 | 0 | 0.67 | 37 | 00000000:06:00.0 | 0 | 3.08 | 100 | 00000000:06:00.0 | 50.45 | 50.45 | 50.45 |
| | | | | 00000000:16:00.0 | 0 | 0.63 | 37 | 00000000:16:00.0 | 0 | 3.09 | 100 | 00000000:16:00.0 | 50.45 | 50.46 | 50.46 |
| | | | | 00000000:96:00.0 | 0 | 0.80 | 38 | 00000000:96:00.0 | 0 | 7.12 | 100 | 00000000:96:00.0 | 50.44 | 50.44 | 50.44 |
| | | | | 00000000:E6:00.0 | 0 | 1.07 | 38 | 00000000:E6:00.0 | 0 | 5.86 | 100 | 00000000:E6:00.0 | 50.43 | 50.43 | 50.43 |
| gpu1202 | 544.71 | 599.46 | 767.75 | 00000000:06:00.0 | 0 | 0.80 | 37 | 00000000:06:00.0 | 0 | 4.63 | 100 | 00000000:06:00.0 | 50.45 | 50.45 | 50.45 |
| | | | | 00000000:16:00.0 | 0 | 0.74 | 38 | 00000000:16:00.0 | 0 | 4.20 | 100 | 00000000:16:00.0 | 50.45 | 50.46 | 50.46 |
| | | | | 00000000:96:00.0 | 0 | 0.71 | 37 | 00000000:96:00.0 | 0 | 4.14 | 100 | 00000000:96:00.0 | 50.44 | 50.44 | 50.44 |
| | | | | 00000000:E6:00.0 | 0 | 0.69 | 37 | 00000000:E6:00.0 | 0 | 4.00 | 100 | 00000000:E6:00.0 | 50.43 | 50.43 | 50.43 |

**Comment by sapkotaruz11** [comment]
@sshivam95  Thanks for the update. I think this is caused by the materialisation of the whole embedding tables into a single model. We could fix this by saving the partial models sequentially without keeping everything in memory at once.

**Comment by sshivam95** [comment]
Can we work on the materialising the model which @sapkotaruz11 talked about in the pipeline after training? I have a implemented a work around which used the sharded checkpoints to create the model but it works outside the pipeline, i.e., I have to explicitly run it on the cpu and it looses the runtime metric of the pipeline.
